### PR TITLE
Fix summarizer nested class bug

### PR DESCRIPTION
**Pull request by mahesh-hegde**
Changes

- [x] always pull in nested classes when a class name is specified.
closes: #303 

- [x] Save all logs to a file in ~/.dart_tool/jnigen/logs/ (Even on success, like CMake does).
closes: #273

- [x] Use java.util.logging instead of printf in `ApiSummarizer`

TODO:

- [x] If config specifies a nested class containing `$`, detect it and fail with an error.
- [x] Test edge cases - Do anonymous classes usage causes problems?

To Discuss:
- [ ] Should we log to file in tests?
- [ ] Should we keep the log file upon success?

---

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by mahesh-hegde** [comment]
```
+              ffi
+              .NativeFunction<jni.JniResult Function(ffi.Pointer<ffi.Void>)>>(
+          "MyStack__pop")
```

Who formats like this? ☠️☠️

**Comment by dcharkes** [comment]
> Should we log to file in tests?

Less file access is probably better. However, we're probably accessing a ton of files in tests already, so a log file doesn't matter that much.

> Should we keep the log file upon success?

What do other projects do?

How do we clean log files? Or are we just going to have a million files on the users' system at some point. (This is relevant both for runs with and without errors.) Should logs older than 30 days automatically be thrown away? Or a week? Or a day?

I'll leave the code review for @HosseinYousefi.

Thanks @mahesh-hegde ! 🚀

**Comment by mahesh-hegde** [comment]
> What do other projects do?

CMake keeps logs on success as well.

> How to clean up.

Everytime before we create new log file, delete every log file except last N files. (Basically log rotation). That's implemented in this PR.

**Comment by mahesh-hegde** [comment]
> ConorMacBride/install-package@3e7ad059e07782ee54fa35f827df52aae0626f30

Do we need to update this action? It may be because homebrew did some changes to their package information format recently.

**Comment by dcharkes** [comment]
> > How to clean up.
> 
> Everytime before we create new log file, delete every log file except last N files. (Basically log rotation). That's implemented in this PR.

Awesome! I'm just asking questions and you've already thought of everything! 🚀

**Comment by HosseinYousefi** [comment]
> > [ConorMacBride/install-package@3e7ad05](https://github.com/ConorMacBride/install-package/commit/3e7ad059e07782ee54fa35f827df52aae0626f30)
> 
> Do we need to update this action? It may be because homebrew did some changes to their package information format recently.

I'll do the CI fixes in another PR.

**Comment by HosseinYousefi** [comment]
Closes #299 too then!

File: jnigen/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+## 0.7.0
```
Comment (HosseinYousefi) [inline jnigen/CHANGELOG.md:1]: I have not published 0.6.0 yet, so this can be in `0.6.0-dev.1` also in pubspec.yaml.

The way I do these change-logs is I start by writing `0.x.0-dev.0` and if it exists I add the next dev version `0.x.0-dev.1`, add the bullet points, and so on... (I also change the version in pubspec) The moment I want to publish I merge all of the `-dev.y`s, remove the `-dev.y` suffix from pubspec and change-logs and I publish. Of course the next breaking change now becomes `0.x+1.0-dev.0` again. 

Perhaps once we setup automatic publishing we can make this process better.

File: jnigen/test/simple_package_test/java/com/github/dart_lang/jnigen/simple_package/Example.java [hunk 2]
```diff
@@ -176,12 +176,27 @@ public Nested(boolean value) {
       this.value = value;
     }
 
+    public void usesAnonymousInnerClass() {
```
Comment (HosseinYousefi) [inline jnigen/test/simple_package_test/java/com/github/dart_lang/jnigen/simple_package/Example.java:179]: Any reason to add this method? It seems not to be tested.

File: jnigen/test/summary_generation_test.dart [hunk 3]
```diff
@@ -21,13 +21,21 @@ import 'package:test/test.dart';
 
 import 'test_util/test_util.dart';
 
+const nestedClasses = [
```
Comment (HosseinYousefi) [inline jnigen/test/summary_generation_test.dart:24]: There are some more nested classes in generics/GrandParent.java.

File: jnigen/pubspec.yaml [hunk 4]
```diff
@@ -3,7 +3,7 @@
 # BSD-style license that can be found in the LICENSE file.
 
 name: jnigen
-version: 0.6.0-dev.0
```
Comment (HosseinYousefi) [inline jnigen/pubspec.yaml:6]: ditto

File: jnigen/test/simple_package_test/java/com/github/dart_lang/jnigen/simple_package/Example.java (hunk 2 above)
Comment (mahesh-hegde) [reply jnigen/test/simple_package_test/java/com/github/dart_lang/jnigen/simple_package/Example.java:179]: It uses an anonymous inner class, which, when compiled, creates a file `Example$Nested$1.class`.

The presence of this class in test ensures presence of anonymous classes don't break code gen.

File: jnigen/test/summary_generation_test.dart (hunk 3 above)
Comment (mahesh-hegde) [reply jnigen/test/summary_generation_test.dart:24]: Done. (This is just verifying nested classes are there. So no need to be exhaustive though).

**Review by HosseinYousefi — APPROVED** [review]

**Comment by HosseinYousefi** [comment]
Thanks @mahesh-hegde!
